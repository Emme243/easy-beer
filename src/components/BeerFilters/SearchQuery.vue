<template>
  <v-text-field
    id="search-query-input"
    :append-icon="searchQuery && 'fa fa-times'"
    :value="searchQuery"
    @click:append="handleSearchQueryChange('')"
    @input="handleSearchQueryChange"
    placeholder="BÃºsqueda de cerveza por nombre"
    prepend-icon="fa-solid fa-magnifying-glass"
  />
</template>

<script>
import { mapActions, mapGetters } from 'vuex';

export default {
  name: 'SearchQuery',
  methods: {
    ...mapActions({
      setSearchQueryInStore: 'filter/setSearchQuery',
      resetPageInStore: 'filter/resetPage',
    }),
    handleSearchQueryChange(searchQuery) {
      this.setSearchQueryInStore(searchQuery);
      this.resetPageInStore();
    },
  },
  computed: mapGetters({ searchQuery: 'filter/searchQuery' }),
};
</script>
