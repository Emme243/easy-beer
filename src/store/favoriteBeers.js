export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
