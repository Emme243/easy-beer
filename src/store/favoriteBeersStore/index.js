export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
