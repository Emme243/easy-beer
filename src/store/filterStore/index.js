import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
