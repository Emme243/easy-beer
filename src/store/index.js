import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
