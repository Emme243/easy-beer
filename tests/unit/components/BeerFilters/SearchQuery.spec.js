import SearchQuery from '@/components/BeerFilters/SearchQuery';
import store from '@/store';
import { createLocalVue, mount } from '@vue/test-utils';
import Vuetify from 'vuetify';

describe('Pruebas en el componente SearchQuery.vue', function () {
  let vuetify;
  const localVue = createLocalVue();

  const mountFunction = () => {
    return mount(SearchQuery, {
      localVue,
      vuetify,
      store,
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
    store.dispatch = jest.fn();
  });

  const newSearchQuery = 'cervecita';
  const textInputSelector = '#search-query-input';

  it('Debe hacer match con el snapshot', function () {
    const wrapper = mountFunction();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Debe llamar al dispatch setSearchQueryInStore y al dispatch resetPage cuando se escribe en el input', async function () {
    const wrapper = mountFunction();
    const inputElement = wrapper.find(textInputSelector);
    await inputElement.setValue(newSearchQuery);
    expect(store.dispatch).toHaveBeenCalledWith('filter/setSearchQuery', newSearchQuery);
    expect(store.dispatch).toHaveBeenCalledWith('filter/resetPage');
    expect(store.dispatch).toHaveBeenCalledTimes(2);
  });
});
