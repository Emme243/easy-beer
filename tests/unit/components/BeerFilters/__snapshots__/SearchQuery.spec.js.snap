// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente SearchQuery.vue Debe hacer match con el snapshot 1`] = `
<div class="v-input theme--light v-text-field v-text-field--placeholder">
  <div class="v-input__prepend-outer">
    <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate fa fa-solid fa-magnifying-glass theme--light"></i></div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot"><input id="search-query-input" placeholder="BÃºsqueda de cerveza por nombre" type="text"></div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
      </div>
    </div>
  </div>
</div>
`;
